### ISSUE TITLE
Bug related to HOME environment variable, Microsoft.Build.Utilities.ToolLocationHelper, and Microsoft.Cpp.x64.user.props or Microsoft.Cpp.Win32.user.props.

### ISSUE THREAD

**Issue by BenKey** [Caphyon/clang-power-tools#542]
I have discovered three bugs in Clang Power Tools related to the HOME environment variable, Microsoft.Build.Utilities.ToolLocationHelper, and Microsoft.Cpp.x64.user.props or Microsoft.Cpp.Win32.user.props.

I discovered the bugs using my [YekNeb project](https://svn.sullivanandkey.com/SnKOpen/cpp/yekneb/trunk/)..

1. If the HOME environment variable is set, the following error occurs.

```
PROJECT: ${ProjectPath}
Set-Variable : Cannot overwrite variable HOME because it is read-only or constant.
At ...\msbuild-project-load.ps1:73 char:7
+       Set-Variable -name $name -Value $value -Scope Global
+       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : WriteError: (HOME:String) [Set-Variable], SessionStateUnauthorizedAccessException
    + FullyQualifiedErrorId : VariableNotWritable,Microsoft.PowerShell.Commands.SetVariableCommand

```

2. If Microsoft.Build.Utilities.ToolLocationHelper is used in a project file, it causes an exception.

My [YekNeb project](https://svn.sullivanandkey.com/SnKOpen/cpp/yekneb/trunk/) uses [Globals.props](https://svn.sullivanandkey.com/SnKOpen/cpp/Shared/trunk/PropertySheets/Globals.props). Globals.props uses the following construct to determine latest installed version of the Windows 10 SDK.

```
<LatestTargetPlatformVersion Condition="'$(VisualStudioVersion)'=='14.0' OR '$(VisualStudioVersion)'=='15.0'">
$([Microsoft.Build.Utilities.ToolLocationHelper]::GetLatestSDKTargetPlatformVersion('Windows', '10.0'))
</LatestTargetPlatformVersion>
```

This construct works correctly in the context of Visual Studio and MSBuild but it causes the following exception with Clang Power Tools.

```
Exception calling "ExpandString" with "1" argument(s): "Unable to find type [Microsoft.Build.Utilities.ToolLocationHelper]."
At ...\msbuild-expression-eval.ps1:166 char:9
+         $res = $ExecutionContext.InvokeCommand.ExpandString($expressi ...
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : RuntimeException
```

3. A number of my projects set additional include directories using Microsoft.Cpp.x64.user.props or Microsoft.Cpp.Win32.user.props which is found in $(LOCALAPPDATA)\Microsoft\MSBuild\v4.0. MSBuild loads settings from these files. This is a common technique for setting user include directories that should be used with all projects.

This technique is documented at [How to include VC++ Directory paths defined in Microsoft,Cpp.User.prop file in MSBuild command line batch file build called in Hudson](https://social.msdn.microsoft.com/Forums/vstudio/en-US/2b2e070c-2918-49ed-a64e-7ef413272e98/how-to-include-vc-directory-paths-defined-in-microsoftcppuserprop-file-in-msbuild-command-line?forum=msbuild) and in various other places.

The Clang Power Tools seem to be unaware of the include directories set using this technique. This causes the following error when I build my YekNeb project.

`error: 'boost/predef.h' file not found [clang-diagnostic-error]
`

The property sheets are part of my [SDKs project](https://svn.sullivanandkey.com/SnKOpen/cpp/SDKs/trunk/) which sets up a standardized build environment for all of my personal projects.

My files can be found in the following locations.

* [Microsoft.Cpp.Win32.user.props](https://svn.sullivanandkey.com/SnKOpen/cpp/SDKs/trunk/props/Microsoft.Cpp.Win32.user.props)
* [Microsoft.Cpp.x64.user.props](https://svn.sullivanandkey.com/SnKOpen/cpp/SDKs/trunk/props/Microsoft.Cpp.x64.user.props)

**Comment by hero101111** [comment]
Hi @BenKey ,

Issue number 3 is already fixed; it'll become available in the next release (`4.9`).
We'll investigate issues 1 & 2 and update this thread as soon as progress is made.

Thank you for taking the time to investigate and document these issues.
Regards,
Gabriel

**Comment by hero101111** [comment]
Hi again,

I'm happy to inform you that we've pushed fixes for all the CPT issues you've encountered while working on your project. They'll be available in the next release (`v4.9`).

Many thanks for the feedback and detailed info.

Regards,
Gabriel.

**Comment by BenKey** [comment]
Hello,

I have installed the latest update. It is working great. Now that I can use
Clang-Tidy on my projects in Windows, I am able to continue my efforts to
improve the quality of my code. I very much appreciate your efforts.
